src/Microsoft.AspNetCore.Mvc.DataAnnotations/Internal/DataAnnotationsModelValidator.cs

```diff
@@ -94,31 +94,52 @@ public IEnumerable<ModelValidationResult> Validate(ModelValidationContext valida
             if (result != ValidationResult.Success)
             {
-                // ModelValidationResult.MemberName is used by invoking validators (such as ModelValidator) to
-                // construct the ModelKey for ModelStateDictionary. When validating at type level we want to append
-                // the returned MemberNames if specified (e.g. person.Address.FirstName). For property validation, the
-                // ModelKey can be constructed using the ModelMetadata and we should ignore MemberName (we don't want
-                // (person.Name.Name). However the invoking validator does not have a way to distinguish between these
-                // two cases. Consequently we'll only set MemberName if this validation returns a MemberName that is
-                // different from the property being validated.
-
-                var errorMemberName = result.MemberNames.FirstOrDefault();
-                if (string.Equals(errorMemberName, memberName, StringComparison.Ordinal))
-                {
-                    errorMemberName = null;
-                }
-
-                string errorMessage = null;
+                string errorMessage;
                 if (_stringLocalizer != null &&
                     !string.IsNullOrEmpty(Attribute.ErrorMessage) &&
                     string.IsNullOrEmpty(Attribute.ErrorMessageResourceName) &&
                     Attribute.ErrorMessageResourceType == null)
                 {
-                    errorMessage = GetErrorMessage(validationContext);
+                    errorMessage = GetErrorMessage(validationContext) ?? result.ErrorMessage;
+                }
+                else
+                {
+                    errorMessage = result.ErrorMessage;
+                }
+
+                var validationResults = new List<ModelValidationResult>();
+                if (result.MemberNames != null)
+                {
+                    foreach (var resultMemberName in result.MemberNames)
+                    {
+                        // ModelValidationResult.MemberName is used by invoking validators (such as ModelValidator) to
+                        // append construct the ModelKey for ModelStateDictionary. When validating at type level we
+                        // want the returned MemberNames if specified (e.g. "person.Address.FirstName"). For property
+                        // validation, the ModelKey can be constructed using the ModelMetadata and we should ignore
+                        // MemberName (we don't want "person.Name.Name"). However the invoking validator does not have
+                        // a way to distinguish between these two cases. Consequently we'll only set MemberName if this
+                        // validation returns a MemberName that is different from the property being validated.
+                        ModelValidationResult validationResult;
+                        if (string.Equals(resultMemberName, memberName, StringComparison.Ordinal))
```

Comment:
Should this be `OrdinalIgnoreCase`? I think it's probably right as is, but I wanted to mention it since lots of stuff is randomly ignore case. Does the data annotations have any docs that might say something about this matter?